test/jit/test_freezing.py

```diff
@@ -0,0 +1,602 @@
+                self.b = 4
+
+            def forward(self, x):
+                self.b = 20
+                return self.sub1(x) + self.a + self.b + self.sub2(x)
+
+        m = torch.jit.script(TestModule())
+        m.eval()
+        input = torch.randn(2, 2)
+        output_s = m.forward(input)
+        mf = torch._C._freeze_module(m._c)
+        self.assertFalse(mf.hasattr('sub1'))
+        self.assertFalse(mf.hasattr('a'))
+        self.assertTrue(mf.hasattr('b'))
+        self.assertTrue(mf.hasattr('sub2'))
+        output_f = mf.forward(input)
+        self.assertEqual(output_s, output_f)
+
+    def test_freeze_module_with_helperfunction(self):
+        class SubModule(nn.Module):
+            def __init__(self):
+                super(SubModule, self).__init__()
+                self.a = 11
+                self.b = 2
+
+            def forward(self, x):
+                return self.a + self.b
+
+        class TestModule(nn.Module):
+            def __init__(self):
+                super(TestModule, self).__init__()
+                self.sub = SubModule()
+                self.a = 3
+                self.b = 4
+
+            def forward(self, x):
+                self.b = 20
+                return self._forward(x) + self.a + self.b
+
+            def _forward(self, x):
+                return self.sub(x)
+        m = torch.jit.script(TestModule())
+        m.eval()
+        input = torch.randn(2, 2)
+        mf = torch._C._freeze_module(m._c)
+        self.assertFalse(mf.hasattr('sub'))
+        self.assertFalse(mf.hasattr('a'))
+        self.assertTrue(mf.hasattr('b'))
+        with self.assertRaisesRegex(RuntimeError, "TestModule does not have a field with name '_forward'"):
+            mf._forward(x)
+
+    def test_freeze_module_with_inplace_mutable(self):
+        class FreezeMe(torch.jit.ScriptModule):
+            def __init__(self):
+                super(FreezeMe, self).__init__()
+                self.a = [11, 22]
+
+            @torch.jit.script_method
+            def forward(self, x):
+                for i in range(3):
+                    self.a.append(i)
+                return self.a
+
+        m = FreezeMe()
+        m.eval()
+        m_f = torch._C._freeze_module(m._c)
+        self.assertTrue(m_f.hasattr('a'))
+        m.forward(torch.tensor([3]))
+        out = m_f.forward(torch.tensor([5]))
+        expected = [11, 22, 0, 1, 2, 0, 1, 2]
+        self.assertEqual(out, expected)
+
+    # Mutable attributes
+    def test_freeze_module_with_mutable_list(self):
+        class FreezeMe(nn.Module):
+            def __init__(self):
+                super(FreezeMe, self).__init__()
+                self.a = [1, 2]
+
+            def forward(self, x):
+                return self.a
+
+        m = FreezeMe()
+        m.eval()
+        m.a.append(3)
+        m_s = torch.jit.script(m)
+        v = m_s.a
+        v.append(4)
+        m_s.a = v
+        m_s.eval()
+        m_f = torch._C._freeze_module(m_s._c)
```

Comment:
Also, before this pass is made public we need a workflow that does not involve accessing private members (`torch._C` and `m_s._c`). This pass should be in `torch.jit.script` with proper documentation (e.g. https://github.com/pytorch/pytorch/blob/master/torch/jit/__init__.py#L1117).